File: go/apib/sender.go
package apib

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"net/http"
	"net/http/httputil"
	"net/url"
	"os"
	"time"
)

const defaultTimeout = 60 * time.Second

type Sender struct {
	client  *http.Client
	url     *url.URL
	urlStr  string
	method  string
	verbose bool
}

func NewSender(urlStr string, expectedConnections int) (*Sender, error) {
	u, err := url.Parse(urlStr)
	if err != nil {
		return nil, fmt.Errorf("invalid URL: %q: %w", urlStr, err)
	}
	return &Sender{
		client: &http.Client{
			Transport: &http.Transport{
				MaxIdleConns:        expectedConnections * 2,
				MaxIdleConnsPerHost: expectedConnections * 2,
				MaxConnsPerHost:     expectedConnections * 2,
			},
		},
		url:    u,
		urlStr: urlStr,
		method: "GET",
	}, nil
}

func (s *Sender) SetVerbose(verbose bool) {
	s.verbose = verbose
}

func (s *Sender) SetMethod(method string) {
	s.method = method
}

func (s *Sender) Send(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, s.method, s.urlStr, &bytes.Buffer{})
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}
	if s.verbose {
		dump, err := httputil.DumpRequestOut(req, false)
		if err != nil {
			return fmt.Errorf("error on dump: %w", err)
		}
		os.Stdout.Write(dump)
	}
	resp, err := s.client.Do(req)
	if err != nil {
		ret := fmt.Errorf("request error: %w", err)
		if s.verbose {
			fmt.Printf("%v\n", ret)
		}
		return ret
	}
	defer resp.Body.Close()
	if s.verbose {
		dump, err := httputil.DumpResponse(resp, false)
		if err != nil {
			return fmt.Errorf("error on dump: %w", err)
		}
		os.Stdout.Write(dump)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("received error status: %d", resp.StatusCode)
	}
	tmp := make([]byte, 8192)
	bytesRead := 0
	for err == nil {
		var r int
		r, err = resp.Body.Read(tmp)
		bytesRead += r
	}
	if err != io.EOF {
		ret := fmt.Errorf("received error %w after reading %d bytes", err, bytesRead)
		if s.verbose {
			fmt.Printf("%v\n", ret)
		}
		return ret
	}
	if s.verbose {
		fmt.Printf("Response body: %d bytes\n", bytesRead)
	}
	return nil
}

func (s *Sender) Loop(ctx context.Context, c *Collector) {
	pleaseStop := false
	for !pleaseStop {
		reqCtx, cancel := context.WithTimeout(ctx, defaultTimeout)
		defer cancel()
		start := time.Now()
		err := s.Send(reqCtx)
		if err == nil {
			pleaseStop = c.Success(start, 0, 0)
		} else {
			pleaseStop = c.Failure(err)
		}
	}
}

File: rust/apib/src/sender.rs
use crate::{
    collector::{Collector, LocalCollector},
    config::Config,
    connector::{Connection, Http1Connection, Http2Connection},
    error::Error,
};
use async_trait::async_trait;
use http_body_util::{BodyExt, Full};
use hyper::{body::Bytes, Request};
use rustls_pki_types::ServerName;
use std::{sync::Arc, time::SystemTime};
use tokio::net::TcpStream;
use tokio_rustls::TlsConnector;

const USER_AGENT: &str = "apib";

#[async_trait]
pub trait Sender {
    async fn send(&mut self) -> Result<bool, Error>;
    async fn do_loop(&mut self, collector: &Collector);
}

struct SenderImpl<C> {
    config: Arc<Config>,
    connection: C,
    request: Option<Request<Full<Bytes>>>,
    verbose: bool,
}

impl<C> SenderImpl<C>
where
    C: Connection + Send,
{
    pub fn new(config: Arc<Config>, connection: C) -> Self {
        let verbose = config.verbose;
        Self {
            config,
            connection,
            request: None,
            verbose,
        }
    }
}

#[async_trait]
impl<C> Sender for SenderImpl<C>
where
    C: Connection + Send,
{
    async fn send(&mut self) -> Result<bool, Error> {
        let mut connection_opened = false;
        if !self.connection.connected() {
            if self.verbose {
                println!("Connecting to {}:{}...", self.config.host, self.config.port);
            }
            let new_conn =
                TcpStream::connect((self.config.host.as_str(), self.config.port)).await?;
            new_conn.set_nodelay(true)?;
            if self.verbose {
                println!("Connected");
            }
            connection_opened = true;

            if let Some(tls_config) = &self.config.tls {
                if self.verbose {
                    println!("Connecting using TLS...");
                }
                let cfg = Arc::clone(tls_config);
                let sn = ServerName::try_from(self.config.host.as_str())
                    .expect("Invalid SNI host name")
                    .to_owned();
                let connector = TlsConnector::from(cfg);
                let tls_conn = connector.connect(sn, new_conn).await?;
                if self.verbose {
                    let (_, tls_info) = tls_conn.get_ref();
                    if let Some(ap) = tls_info.alpn_protocol() {
                        println!("ALPN protocol: {:?}", ap);
                    }
                    if let Some(pv) = tls_info.protocol_version() {
                        println!("Connected using {:?}", pv);
                    }
                    if let Some(cs) = tls_info.negotiated_cipher_suite() {
                        println!("Cipher: {:?}", cs);
                    }
                }
                self.connection.connect(tls_conn).await?
            } else {
                self.connection.connect(new_conn).await?
            }
        }

        let request = match &self.request {
            Some(r) => r.clone(),
            None => {
                let mut req_builder = Request::builder()
                    .uri(self.config.path.as_str())
                    .method(&self.config.method)
                    .header("Host", self.config.host_hdr.as_str())
                    .header("User-Agent", USER_AGENT);
                for (name, value) in &self.config.headers {
                    req_builder = req_builder.header(name, value);
                }
                let new_req = req_builder.body(Full::new(self.config.body.clone()))?;
                self.request = Some(new_req.clone());
                new_req
            }
        };
        if self.verbose {
            println!("{:?}", request);
        }

        let mut response = match self.connection.send_request(request).await {
            Ok(r) => r,
            Err(e) => {
                self.connection.disconnect();
                return Err(e);
            }
        };

        if let Some(close_header) = response.headers().get("close") {
            if close_header == "close" {
                self.connection.disconnect();
            }
        }

        if !response.status().is_success() {
            //  Don't deliberately disconnect now
            return Err(Error::Http(response.status().as_u16()));
        }
        if self.verbose {
            for (key, value) in response.headers().iter() {
                println!("{}: {}", key, value.to_str().unwrap());
            }
            let body_buf = response.body_mut().collect().await?.to_bytes();
            println!("\n{:?}", body_buf);
        } else {
            while response.frame().await.is_some() {}
        }
        Ok(connection_opened)
    }

    async fn do_loop(&mut self, collector: &Collector) {
        let mut local_stats = LocalCollector::new();
        loop {
            let start = SystemTime::now();
            match self.send().await {
                Ok(connection_opened) => {
                    if connection_opened {
                        Collector::connection_opened(&mut local_stats);
                    }
                    if collector.success(&mut local_stats, start, SystemTime::now(), 0, 0) {
                        break;
                    }
                }
                Err(e) => {
                    if self.verbose {
                        println!("Error: {}", e);
                    }
                    if collector.failure(&mut local_stats, e) {
                        break;
                    }
                }
            }
        }
        collector.collect(local_stats);
    }
}

// Create a new sender. HTTP/1 and 2 use different implementations, and
// there are a bunch of trait-related things that happen in the background,
// so this function abstracts that away.
pub fn new_sender(config: Arc<Config>, http2: bool) -> Box<dyn Sender + Send> {
    if http2 {
        Box::new(SenderImpl::new(config, Http2Connection::new()))
    } else {
        Box::new(SenderImpl::new(config, Http1Connection::new()))
    }
}
